state-commitments/optimized/history_commitment.go

```diff
@@ -50,18 +77,43 @@ func NewCommitter() *HistoryCommitter {
 	}
 }
 
+func (h *HistoryCommitter) hash(item []byte) (common.Hash, error) {
+	defer h.keccak.Reset()
```

Comment:
Is reusing the crypto.KeccakState here really significantly faster than just using `crypto.Keccak256Hash(item)` (assuming we import geth as a dependency)? Have we benchmarked the two?

Also, if we do keep the current impl, I'd just panic on a read/write error in this function. It should never happen, and I think the program dying in that case makes a lot of sense.

Reply:
So, it's about 12.9% faster in a tight benchmark.

```
❯ go test ./state-commitments/history -bench="BenchmarkHashFunction|BenchmarkKeccak256Hash" -run="^$" -benchtime=20s
goos: darwin
goarch: arm64
pkg: github.com/OffchainLabs/bold/state-commitments/history
BenchmarkHashFunction-12        76132399               292.1 ns/op
BenchmarkKeccak256Hash-12       72502815               335.4 ns/op
PASS
ok      github.com/OffchainLabs/bold/state-commitments/history  47.498s

❯ echo "(335.4 - 292.1) / 335.4" | bc -l
.12909958258795468097
```

I think that justifies doing it this way.
I also agree that having this function return an error is not really useful.
In fact, I'm tempted to do what the `crypto` library itself does and just blindly ignore the errors (rather than bothering to panic.) I'm updating the code to do that. If you really want it to panic, reopen this conversation.